### PR TITLE
fix: crash on publishing role to non-publishing role and back

### PR DESCRIPTION
**Pull request by ygit**
### Details(context, link the issue, how was the bug fixed, what does the new feature do)

1. Corrected typo in device label
2. Added checks for the currently selected audio output
3. Corrected check for device categories

### CODE REVIEW

File: packages/hms-video-store/src/device-manager/DeviceManager.ts [hunk 1]
```diff
@@ -450,7 +450,7 @@ export class DeviceManager implements HMSDeviceManager {
         wired = device;
       } else if (deviceCategory === 'bluetooth') {
         bluetoothDevice = device;
-      } else if (deviceCategory === 'speakerhone') {
+      } else if (deviceCategory === 'earpiece') {
```
Comment (raviteja83) [inline packages/hms-video-store/src/device-manager/DeviceManager.ts:453]: @eswarclynn could this typo have caused any issues?

File: packages/roomkit-react/src/Prebuilt/components/Header/common.jsx [hunk 2]
```diff
@@ -79,7 +79,7 @@ export const AudioActions = () => {
   if (!audioFiltered) {
     return null;
   }
-  const deviceCategory = getAudioDeviceCategory(currentSelection.label);
+  let deviceCategory = getAudioDeviceCategory(currentSelection?.label);
```
Comment (raviteja83) [inline packages/roomkit-react/src/Prebuilt/components/Header/common.jsx:82]: can be const only no? are we re-assigning?

File: packages/hms-video-store/src/utils/media.ts [hunk 3]
```diff
@@ -64,10 +64,14 @@ export const HMSAudioContextHandler: HMSAudioContext = {
   },
 };
 
-export const getAudioDeviceCategory = (deviceLabel: string) => {
+export const getAudioDeviceCategory = (deviceLabel?: string) => {
+  if (!deviceLabel) {
+    HMSLogger.e('[DeviceManager]:', 'No device label provided');
+    return 'speakerphone';
```
Comment (raviteja83) [inline packages/hms-video-store/src/utils/media.ts:70]: when do we get into this scenario? when no publish permission is present like viewer roles?

File: packages/hms-video-store/src/utils/media.ts [hunk 4]
```diff
@@ -64,10 +64,14 @@ export const HMSAudioContextHandler: HMSAudioContext = {
...
+  }
   const label = deviceLabel.toLowerCase();
   if (label.includes('speakerphone')) {
-    return 'speakerhone';
+    return 'speakerphone';
```
Comment (raviteja83) [inline packages/hms-video-store/src/utils/media.ts:74]: @eswarclynn enum could have prevented this.

File: packages/hms-video-store/src/device-manager/DeviceManager.ts (hunk 1 above)
Comment (raviteja83) [reply packages/hms-video-store/src/device-manager/DeviceManager.ts:453]: our earpiece selection logic will fail

File: packages/roomkit-react/src/Prebuilt/components/Header/common.jsx (hunk 2 above)
Comment (ygit) [reply packages/roomkit-react/src/Prebuilt/components/Header/common.jsx:82]: yes, can be `const`

File: packages/hms-video-store/src/utils/media.ts (hunk 3 above)
Comment (ygit) [reply packages/hms-video-store/src/utils/media.ts:70]: 1. In LiveStreaming Room, join as Broadcaster
2. Switch role to VNRT
3. Switch role back to Broadcaster, then the crash happens

File: packages/hms-video-store/src/device-manager/DeviceManager.ts (hunk 1 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/device-manager/DeviceManager.ts:453]: Could be remember copying this code, from AudioSinkManager

File: packages/hms-video-store/src/utils/media.ts (hunk 4 above)
Comment (eswarclynn) [reply packages/hms-video-store/src/utils/media.ts:74]: ✅

**Review by KaustubhKumar05 — DISMISSED** [review]

File: packages/hms-video-store/src/utils/media.ts (hunk 4 above)
Comment (hdz-666) [reply packages/hms-video-store/src/utils/media.ts:74]: Lets use one ? @ygit

Comment (ygit) [reply packages/hms-video-store/src/utils/media.ts:74]: @hdz-666 Will create enum in another PR

File: packages/roomkit-react/src/Prebuilt/components/Header/common.jsx [hunk 5]
```diff
@@ -25,6 +25,7 @@ import { Sheet } from '../../../Sheet';
 import { Text } from '../../../Text';
 import IconButton from '../../IconButton';
 import { ToastManager } from '../Toast/ToastManager';
+import { AudioDeviceCategory } from '@100mslive/hms-video-store/src/utils/media';
```
Comment (raviteja83) [inline packages/roomkit-react/src/Prebuilt/components/Header/common.jsx:28]: it should not be imported like this.

File: packages/hms-video-store/src/utils/media.ts [hunk 6]
```diff
@@ -64,20 +64,27 @@ export const HMSAudioContextHandler: HMSAudioContext = {
   },
 };
 
+export enum AudioDeviceCategory {
```
Comment (raviteja83) [inline packages/hms-video-store/src/utils/media.ts:67]: this needs to be exported again if not already exported to be used roomkit-react

Comment (KaustubhKumar05) [reply packages/hms-video-store/src/utils/media.ts:67]: @ygit you can refer HMSAudioMode or HMSNotificationTypes to fix the import and exports

File: packages/roomkit-react/src/Prebuilt/components/Header/common.jsx [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+import { HMSAudioDeviceCategory } from '@100mslive/hms-video-store';
```
Comment (raviteja83) [inline packages/roomkit-react/src/Prebuilt/components/Header/common.jsx:2]: should be from react-sdk. hms-video-store is not a direct dependency

**Review by amar-1995 — DISMISSED** [review]

**Review by raviteja83 — DISMISSED** [review]

File: packages/hms-video-store/src/utils/media.ts [hunk 8]
```diff
@@ -64,16 +64,27 @@ export const HMSAudioContextHandler: HMSAudioContext = {
   },
 };
 
-export const getAudioDeviceCategory = (deviceLabel: string) => {
+export enum HMSAudioDeviceCategory {
+  SPEAKERPHONE,
```
Comment (raviteja83) [inline packages/hms-video-store/src/utils/media.ts:68]: should we use string enums?

Comment (ygit) [reply packages/hms-video-store/src/utils/media.ts:68]: Sure can do

**Review by KaustubhKumar05 — APPROVED** [review]

**Review by raviteja83 — APPROVED** [review]

**Pull request merged**